### PR TITLE
server: fix issues with critical config check

### LINKED ISSUE #12238: v3.0.11 upgrade to v6.0.0 fail with "critical config check failed: raft engine dir have been changed, former is ''
### What version of TiKV are you using?
v6.0.0                                               2022-03-22T19:50:34+08:00 

### Steps to reproduce

install v3.0.11 with default config 
upgrade to v6.0.0
### What did you expect?
upgrade successfully

### What did happened?
upgrade fail with:
Error: init config failed: tikv1-peer:30160: executor.ssh.execute_failed: Failed to execute command over SSH for 'tidb@tikv1-peer:22' {ssh_stderr: critical config check failed: raft engine dir have been changed, former is '', current is '/tiup/data/tikv-30160/raft-engine', please check if it is expected.
, ssh_stdout: , ssh_command: export LANG=C; PATH=$PATH:/bin:/sbin:/usr/bin:/usr/sbin /tiup/deploy/tikv-30160/bin/tikv-server --config-check --config=/tiup/deploy/tikv-30160/conf/tikv.toml --pd "" --data-dir "/tiup/data/tikv-30160"}, cause: Process exited with status 1: check config failed

### PR DESCRIPTION
**Pull request by tabokie**
Signed-off-by: tabokie <xy.tao@outlook.com>

### What is changed and how it works?

Issue Number: Close #12238

What's Changed:

I observed at least two issues, both commented in the patch. The direct cause for #12238 is probably a silent error at this line, haven't verified.
https://github.com/tikv/tikv/blob/5ee7d037d4e2e63d710410daac81baca7e9efe69/src/config.rs#L3185

```commit-message
```

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -3042,24 +3042,6 @@ impl TiKvConfig {
     }
 
     pub fn check_critical_cfg_with(&self, last_cfg: &Self) -> Result<(), String> {
-        if last_cfg.rocksdb.wal_dir != self.rocksdb.wal_dir {
```
Comment (Connor1996) [inline src/config.rs:3045]: you should find the reason, the check looks reasonable

File: src/config.rs [hunk 2]
```diff
@@ -3074,32 +3056,49 @@ impl TiKvConfig {
...
+                 have been moved to destination directory.",
+                last_cfg.rocksdb.wal_dir, self.rocksdb.wal_dir
+            ));
+        }
 
-        if last_cfg.raft_store.raftdb_path != self.raft_store.raftdb_path
-            && !last_cfg.raft_engine.enable
-        {
+        // It's possible that `last_cfg` is not fully validated.
+        let last_raftdb_dir = last_cfg
```
Comment (glorv) [inline src/config.rs:3069]: config.validate should already do this adjust. 

https://github.com/tikv/tikv/blob/5ee7d037d4e2e63d710410daac81baca7e9efe69/src/config.rs#L3183-L3187

https://github.com/tikv/tikv/blob/5ee7d037d4e2e63d710410daac81baca7e9efe69/src/config.rs#L2691-L2692

File: src/config.rs (hunk 1 above)
Comment (tabokie) [reply src/config.rs:3045]: I didn't remove any check.

File: src/config.rs (hunk 2 above)
Comment (tabokie) [reply src/config.rs:3069]: See my PR description. Validating config from v3.0 could possibly fail at this line:
https://github.com/tikv/tikv/blob/5ee7d037d4e2e63d710410daac81baca7e9efe69/src/config.rs#L3185

I didn't verify though.

File: src/config.rs (hunk 1 above)
Comment (tabokie) [reply src/config.rs:3045]: Verified in https://github.com/tikv/tikv/issues/12238.

File: src/config.rs (hunk 2 above)
Comment (you06) [reply src/config.rs:3069]: I'm not sure about the risk of config adjustment skipped by validation error, may warn this error.

Comment (tabokie) [reply src/config.rs:3069]: Alright.

**Review by you06 — APPROVED** [review]

**Review by glorv — APPROVED** [review]

File: src/config.rs [hunk 3]
```diff
@@ -3074,32 +3056,50 @@ impl TiKvConfig {
                 ));
             }
         }
+        if last_cfg.rocksdb.wal_dir != self.rocksdb.wal_dir {
+            return Err(format!(
+                "db wal dir have been changed, former db wal_dir is '{}', \
```
Comment (hunterlxt) [inline src/config.rs:3061]:
```suggestion
                "db wal dir have been changed, former is '{}', \
```

**Review by hunterlxt — APPROVED** [review]
rest LGTM

**Pull request merged**